### PR TITLE
DamageModifierSkill

### PR DESCRIPTION
**Pull request by BotAngus**
Adds the damage modifier skill to the superheroes plugin.

Example

```yaml
truehit:
  skill: DAMAGEMODIFIER
  expectedMaxDamage: 30
  damageMin: 200
  damageMax: 30
  outgoing: true
  incoming: false
  priority: 100
  eased: true
  entitywhitelist: []
  entityblacklist: []
  causeblacklist: []
  causewhitelist: []
  ```

### CODE REVIEW

File: src/main/java/me/xemor/superheroes/skills/implementations/DamageModifierSkill.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    @EventHandler
+    public void onDamage(EntityDamageByEntityEvent event) {
+
+
+        double offence = event.getDamage();
+        double offencePrioity = Integer.MIN_VALUE;
+
+        double defence = event.getDamage();
+        double defencePriority = Integer.MIN_VALUE;
```
Comment (Xemorr) [inline src/main/java/me/xemor/superheroes/skills/implementations/DamageModifierSkill.java:27]: This priority system is clever, but I'm not sure when it would ever be used. Does it default to having the defender have higher priority, ie calculate offense then calculate defense?

File: src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        // whitelist nonempty and it is IN it
+        // both are empty
+        return (!entityblacklist.getSet().isEmpty() && !entityblacklist.inSet(entityType)) ||
+            (!entitywhitelist.getSet().isEmpty() && entitywhitelist.inSet(entityType)) ||
+            (entitywhitelist.getSet().isEmpty() && entityblacklist.getSet()
+                .isEmpty());
+    }
+
+    public final double calculateDamage(double damage) {
+        double x = damage;
```
Comment (Xemorr) [inline src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java:89]: Is this line necessary? It just obfuscates the code for no reason as far as I can tell

File: src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    private final boolean outgoing;
+    private final boolean eased;
+
+    public DamageModifierData(int skill,
+                              @NotNull ConfigurationSection configurationSection) {
+        super(skill, configurationSection);
+
+        expectedMaximumDamage = configurationSection.getDouble("expectedMaxDamage", 30);
+        damageCap = configurationSection.getDouble("damageMax", 15);
+        damageMin = configurationSection.getDouble("damageMin", 0);
```
Comment (Xemorr) [inline src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java:36]: Make your ordering of max and min with word damage consistent in the user facing variables, ie maxDamage and expectedMaxDamage or expectedDamageMax and damageMax (the second ones sound stupid so probs go with first lot)

File: src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        causeblacklist =
+            new SetData<>(EntityDamageEvent.DamageCause.class, "causeblacklist",
+                configurationSection);
+        causewhitelist =
```
Comment (Xemorr) [inline src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java:40]: tbh I'd be tempted to just have "causes" and "entities" then a toggle to swap between blacklist and whitelist, I think this is how I normally do it

File: src/main/java/me/xemor/superheroes/skills/implementations/DamageModifierSkill.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            for (SkillData skillData : skillDataCollection) {
+                DamageModifierData damageModifierData =
+                    (DamageModifierData) skillData;
+
+                if (damageModifierData.isValidCause(event.getCause())) {
+                    double newDamage = damageModifierData.calculateDamage(event.getDamage());
+                    event.setDamage(newDamage);
+
+                }
+
```
Comment (Xemorr) [inline src/main/java/me/xemor/superheroes/skills/implementations/DamageModifierSkill.java:95]: remove these random new lines in the braces

File: src/main/java/me/xemor/superheroes/skills/implementations/DamageModifierSkill.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            for (SkillData skillData : skillDataCollection) {
+                DamageModifierData damageModifierData = (DamageModifierData) skillData;
+
+                if (damageModifierData.isOutgoing() &&
+                    damageModifierData.isValidEntity(event.getEntity().getType())) {
+                    if (damageModifierData.getPriority() > offencePrioity) {
+                        offence = damageModifierData.calculateDamage(event.getDamage());
+                        offencePrioity = damageModifierData.getPriority();
+                    }
+
```
Comment (Xemorr) [inline src/main/java/me/xemor/superheroes/skills/implementations/DamageModifierSkill.java:46]: remove new lines in braces

**Review by Xemorr — CHANGES_REQUESTED** [review]

File: src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java (hunk 4 above)
Comment (Xemorr) [reply src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java:40]: then have blacklist on true by default (assume if they don't specify, it allows all)

File: src/main/java/me/xemor/superheroes/skills/implementations/DamageModifierSkill.java (hunk 1 above)
Comment (BotAngus) [reply src/main/java/me/xemor/superheroes/skills/implementations/DamageModifierSkill.java:27]: It defaults to defender having higher priority. 

As for when it would be used, a possible situation might be a defense oriented hero who is intended to be stronger then a particular offense oriented hero for lore, roleplay or gameplay reasons.

File: src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java (hunk 4 above)
Comment (BotAngus) [reply src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java]: whitelist: false is on by default

File: src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java [hunk 7]
```diff
@@ -8,49 +8,41 @@
...
-        damageMin = configurationSection.getDouble("damageMin", 0);
-        causeblacklist =
-            new SetData<>(EntityDamageEvent.DamageCause.class, "causeblacklist",
-                configurationSection);
-        causewhitelist =
-            new SetData<>(EntityDamageEvent.DamageCause.class, "causewhitelist",
-                configurationSection);
+        expectedMaxDamage = configurationSection.getDouble("expectedmaxdamage", 30);
+        maxDamage = configurationSection.getDouble("maxdamage", 15);
+        minDamage = configurationSection.getDouble("mindamage", 0);
```
Comment (Xemorr) [inline src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java:32]: keep the camel case

File: src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java [hunk 8]
```diff
@@ -69,34 +61,27 @@ public final boolean isValidCause(EntityDamageEvent.DamageCause damageCause) {
...
-        return (!entityblacklist.getSet().isEmpty() && !entityblacklist.inSet(entityType)) ||
-            (!entitywhitelist.getSet().isEmpty() && entitywhitelist.inSet(entityType)) ||
-            (entitywhitelist.getSet().isEmpty() && entityblacklist.getSet()
-                .isEmpty());
+        return (whitelist && !entities.inSet(entityType));
     }
 
-    public final double calculateDamage(double damage) {
-        double x = damage;
+    public final double calculateDamage(double x) {
```
Comment (Xemorr) [inline src/main/java/me/xemor/superheroes/skills/skilldata/DamageModifierData.java:74]: why did you change the variable name to x, that's what I meant by the obfuscation. Use a sensible name
